signer/plugin.go

```diff
@@ -0,0 +1,259 @@
+	"github.com/notaryproject/notation-go/plugin"
+	"github.com/notaryproject/notation-go/plugin/proto"
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+	"oras.land/oras-go/v2/content"
+)
+
+// pluginSigner signs artifacts and generates signatures.
+// It implements notation.Signer
+type pluginSigner struct {
+	plugin       plugin.SignPlugin
+	keyID        string
+	pluginConfig map[string]string
+	signer       signature.Signer
+}
+
+// NewSignerPlugin creates a notation.Signer that signs artifacts and generates signatures
+// by delegating the one or more operations to the named plugin, as defined in
+// https://github.com/notaryproject/notaryproject/blob/main/specs/plugin-extensibility.md#signing-interfaces.
+func NewFromPlugin(plugin plugin.Plugin, keyID string, pluginConfig map[string]string) (notation.Signer, error) {
+	if plugin == nil {
+		return nil, errors.New("nil plugin")
+	}
+	if keyID == "" {
+		return nil, errors.New("nil signing keyID")
+	}
+	// remoteSigner implements signature.Signer
+	remoteSigner := &remoteSigner{
+		plugin:       plugin,
+		pluginConfig: pluginConfig,
+		keyID:        keyID,
+	}
+	return &pluginSigner{
+		plugin:       plugin,
+		keyID:        keyID,
+		pluginConfig: pluginConfig,
+		signer:       remoteSigner,
+	}, nil
+}
+
+// Sign signs the artifact described by its descriptor and returns the marshalled envelope.
+func (s *pluginSigner) Sign(ctx context.Context, desc ocispec.Descriptor, opts notation.SignOptions) ([]byte, *signature.SignerInfo, error) {
+	req := &proto.GetMetadataRequest{
+		PluginConfig: s.pluginConfig,
+	}
+	metadata, err := s.plugin.GetMetadata(ctx, req)
+	if err != nil {
+		return nil, nil, err
+	}
+	if !metadata.SupportsContract(proto.ContractVersion) {
+		return nil, nil, fmt.Errorf(
+			"contract version %q is not in the list of the plugin supported versions %v",
+			proto.ContractVersion, metadata.SupportedContractVersions,
+		)
+	}
+	if metadata.HasCapability(proto.CapabilitySignatureGenerator) {
+		return s.generateSignature(ctx, desc, opts)
+	} else if metadata.HasCapability(proto.CapabilityEnvelopeGenerator) {
+		return s.generateSignatureEnvelope(ctx, desc, opts)
+	}
+	return nil, nil, fmt.Errorf("plugin does not have signing capabilities")
+}
+
+func (s *pluginSigner) generateSignature(ctx context.Context, desc ocispec.Descriptor, opts notation.SignOptions) ([]byte, *signature.SignerInfo, error) {
+	config := s.mergeConfig(opts.PluginConfig)
+	// Get key info.
+	key, err := s.describeKey(ctx, config)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// Check keyID is honored.
+	if s.keyID != key.KeyID {
+		return nil, nil, fmt.Errorf("keyID in describeKey response %q does not match request %q", key.KeyID, s.keyID)
+	}
+	ks, err := proto.DecodeKeySpec(key.KeySpec)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	if remoteSigner, ok := s.signer.(*remoteSigner); ok {
+		remoteSigner.keySpec = ks
+		remoteSigner.pluginConfig = config
+	}
```

Comment:
The `remoteSigner` is a one-time-use signer. It should not be a part of `pluginSigner`. In other words, the field `pluginSigner.Signer` should be removed.

The `remoteSigner` should be created per call of `generateSignature` since the `config` is different.

Reply:
> The `remoteSigner` is a one-time-use signer. It should not be a part of `pluginSigner`. In other words, the field `pluginSigner.Signer` should be removed.
> 
> The `remoteSigner` should be created per call of `generateSignature` since the `config` is different.

The change will make pluginSigner.generateSignature() always use the remoteSigner struct as the implementation of signature.Signer. In other words, upstream callers cannot bring their own implementations of signature.Signer such as mockRemoteSigner in the unit test. Is this expected? (The original logic has this flexibility though.)

Reply:
Providing arbitrary `signature.Signer` to a `pluginSigner` is indeed an instance of built-in signer class.

Let's revisit the design.
```mermaid
classDiagram

class Signer
<<notation>> Signer
Signer: +Sign()

class genericSigner
genericSigner: notation-core-go/signature.Signer
genericSigner: +Sign()

class pluginEnvelopeSigner
pluginEnvelopeSigner: +Sign()

class pluginPrimitiveSigner
pluginPrimitiveSigner: +Sign()
pluginPrimitiveSigner: +KeySpec()

class builtinSigner
builtinSigner: +Sign()
builtinSigner: +KeySpec()

class CLIPlugin

Signer <|-- genericSigner: implements
Signer <|-- pluginEnvelopeSigner: implements

genericSigner o-- builtinSigner
genericSigner o-- pluginPrimitiveSigner

pluginEnvelopeSigner o-- CLIPlugin
pluginPrimitiveSigner o-- CLIPlugin
```

Reply:
Since we determine the plugin signing capability on-the-fly, the above design can be involved into
```mermaid
classDiagram

class Signer
<<notation>> Signer
Signer: +Sign()

class genericSigner
genericSigner: notation-core-go/signature.Signer
genericSigner: +Sign()

class pluginSigner
pluginSigner: +Sign()

class pluginPrimitiveSigner
pluginPrimitiveSigner: +Sign()
pluginPrimitiveSigner: +KeySpec()

class builtinSigner
builtinSigner: +Sign()
builtinSigner: +KeySpec()

class CLIPlugin

Signer <|-- genericSigner: implements
Signer <|-- pluginSigner: implements

genericSigner o-- builtinSigner
genericSigner o-- pluginPrimitiveSigner

pluginSigner o-- CLIPlugin
pluginPrimitiveSigner o-- CLIPlugin

genericSigner .. pluginSigner
```
Depending on the plugin capability, `pluginSigner` can instantiate a `genericSigner` with `pluginPrimitiveSigner`.

Reply:
In fact, you are already implementing the above design with slight differences and a different name set.

```mermaid
classDiagram

class Signer
<<notation>> Signer
Signer: +Sign()

class builtinSigner
builtinSigner: notation-core-go/signature.Signer
builtinSigner: +Sign()

class pluginSigner
pluginSigner: +Sign()

class remoteSigner
remoteSigner: +Sign()
remoteSigner: +KeySpec()

class localSigner
localSigner: +Sign()
localSigner: +KeySpec()

class CLIPlugin

Signer <|-- builtinSigner: implements
Signer <|-- pluginSigner: implements

builtinSigner o-- localSigner
builtinSigner o-- remoteSigner

pluginSigner o-- CLIPlugin
remoteSigner o-- CLIPlugin

builtinSigner .. pluginSigner
```

Reply:
@shizhMSFT So looks like the current design is pretty much the same as your proposed one besides the names? Are you suggesting that I should rename them?